src/dw_warp/dw_warp.c

```diff
@@ -0,0 +1,368 @@
+
+typedef struct DistortionWorldWarp {
+    GenericPointerData * p3DCallback;
+    UnkStruct_020203AC * camera;
+    SysTask * task;
+    int frameCnt;
+    int soundEffectCnt;
+    UnkStruct_02017294 animationObj;
+    UnkStruct_02017248 animationAnimation;
+    UnkStruct_02017248 animationAnimation2;
+    UnkStruct_020170F4 animationModel;
+    u32 moveFrame;
+    int cameraAngle;
+    int cameraAngleCalc;
+    fx32 cameraDistance;
+    int cameraPerspective;
+    NNSFndAllocator allocator;
+} DistortionWorldWarp;
+
+static void DW_Update(SysTask * task, void * data);
+static void DW_VBlankIntr(void * data);
+static void DW_VramSetBank(void);
+static void DW_InitCamera(DistortionWorldWarp * warp);
+static void DWWarp_CameraEnd(DistortionWorldWarp * warp);
+static void DWWarp_InitModel(DistortionWorldWarp * warp);
+static void DWWarp_DeleteModel(DistortionWorldWarp * warp);
+static void Model3D_Update(DistortionWorldWarp * warp);
+static GenericPointerData * DWWarp_Init3D(int heapId);
+static void DWWarp_Setup3D(void);
+static void DWWarp_Exit3D(GenericPointerData * param0);
+static void DWWarp_CameraMove(DistortionWorldWarp * warp);
+
+int DWWarp_Init (OverlayManager * ovy, int * state)
+{
+    SetMainCallback(NULL, NULL);
+    DisableHBlank();
+    GXLayers_DisableEngineALayers();
+    GXLayers_DisableEngineBLayers();
+
+    GX_SetVisiblePlane(0);
+    GXS_SetVisiblePlane(0);
+    GX_SetVisibleWnd(GX_WNDMASK_NONE);
+    GXS_SetVisibleWnd(GX_WNDMASK_NONE);
+    G2_BlendNone();
+    G2S_BlendNone();
+
+    Heap_Create(HEAP_ID_APPLICATION, HEAP_ID_DISTORTION_WORLD_WARP, HEAP_SIZE_DISTORTION_WORLD_WARP);
+
+    DistortionWorldWarp* dww = OverlayManager_NewData(ovy, sizeof(DistortionWorldWarp), HEAP_ID_DISTORTION_WORLD_WARP);
+    MI_CpuClear8(dww, sizeof(DistortionWorldWarp));
+    dww->p3DCallback = DWWarp_Init3D(HEAP_ID_DISTORTION_WORLD_WARP);
+
+    SetAutorepeat(4, 8);
+    DW_VramSetBank();
+    sub_0201E3D8();
+    sub_0201E450(4);
+
+    DWWarp_InitModel(dww);
+    DW_InitCamera(dww);
+    sub_0200F174(0, 1, 1, 0x0, 16, 1, HEAP_ID_DISTORTION_WORLD_WARP);
+
+    gCoreSys.unk_65 = 0;
+
+    GXLayers_SwapDisplay();
+    GXLayers_TurnBothDispOn();
+    sub_02002AC8(1);
+    sub_02002AE4(0);
+    sub_02002B20(0);
+
+    dww->task = SysTask_Start(DW_Update, dww, 60000);
+    SetMainCallback(DW_VBlankIntr, dww);
+
+    return TRUE;
+}
+
+enum {
+    DWARP_SEQ_SCREENWIPE = 0,
+    DWARP_SEQ_LOOP,
+    DWARP_SEQ_CLEAR_SCREEN,
+    DWARP_SEQ_WAIT
+};
+
+int DWWarp_Main (OverlayManager * ovy, int * state)
+{
+    DistortionWorldWarp * warp = OverlayManager_Data(ovy);
+
+    // Handle the different states of the warp effect.
+    switch (*state) {
+    case DWARP_SEQ_SCREENWIPE:
+        if (ScreenWipe_Done() == TRUE) {
+            (*state)++;
+        }
+        break;
+    case DWARP_SEQ_LOOP:
+        // Play the portal sound effect after 15 frames
+        if (warp->soundEffectCnt == 15) {
+            Sound_PlayEffect(1491);
```

Comment:
**polish:** Use a constant from `consts/sdat.h`; ideally, name the constant to an English equivalent, as well.